// night-owl/lib/utils/logger.js
import pc from "picocolors"
import config from "../config/index.js"

const LEAD = pc.blue("[night-owl]")
const FAIL = pc.red("Error")
const WARN = pc.yellow("Warning")

const prefix = () => `${LEAD} ${pc.gray(timestamp())}`

const timestamp = Intl.DateTimeFormat(undefined, {
  hour: "2-digit",
  minute: "2-digit",
  second: "2-digit",
  hourCycle: "h23",
}).format

const delta = (t) => (performance.now() - t).toFixed(3) + "ms"

const log = (...args) => console.log(prefix(), ...args)

export const time = async (label, fn) => {
  if (!config.logger?.time) return fn()

  const t = performance.now()

  return Promise.resolve(fn()).then((res) => {
    log(pc.gray([label, delta(t)].join(" ")))
    return res
  })
}

export const info = (...args) => {
  if (config.logger?.info) {
    args.forEach((arg) => log(arg))
  }
}

export const fail = (label, err) => {
  if (config.logger?.fail) {
    if (err?.message.match("\n")) err = err.message.trim()

    log(FAIL, label, ...(err ? ["\n\n", err] : []), "\n")
  }
}

export const warn = (...args) => {
  if (config.logger?.warn) {
    args.forEach((arg) => log(WARN, arg))
  }
}
